### PR TITLE
James 1760 Add container affinity in docker based tests

### PR DESCRIPTION
**Pull request by aduprat**
Only the last commit has to be reviewed.
The other is in review here: #281

### CODE REVIEW

File: server/container/util-java8/src/test/java/org/apache/james/util/streams/SwarmGenericContainer.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public GenericContainer withAffinityToContainer() {
+        String container = System.getenv(DOCKER_CONTAINER);
+        if (Strings.isNullOrEmpty(container)) {
+            LOGGER.warn("'DOCKER_CONTAINER' environment variable not found, dockering without affinity");
+            return this;
+        }
+        List<String> envVariables = getEnv();
+        envVariables.add("affinity:container==" + container);
```
Comment (rouazana) [inline server/container/util-java8/src/test/java/org/apache/james/util/streams/SwarmGenericContainer.java]: Maybe you can use addEnv() instead.

Comment (aduprat) [inline server/container/util-java8/src/test/java/org/apache/james/util/streams/SwarmGenericContainer.java]: I think we have to distinct docker env variables & swarm env variables handling:
- docker env are KEY=VALUE
- swarm affinity env are affinity:container==ID or affinity:container!=ID

I'm not confortable with the usage of addEnv in such case, which leads to something like that:
```java
addEnv("affinity:container=", "value");
```
as this methods is appending `key + "=" + value` in the env list.

File: README.txt [hunk 2]
```diff
@@ -84,6 +84,10 @@ as it is needed by the container that will run James.
 - SHA1 (optional): is the given git SHA1 of the james-project repository to build or trunk if none.
 - -s option: given tests will not be played while building. Not specifying means play tests.
 
+Some tests needs a DOCKER_HOST environment variable in order to be played, they will be ignored if you don't provide this variable.
+If you wish to play them, you may use a command like the following (depending on your docker configuration):
+$ docker run --env DOCKER_HOST=tcp://172.17.0.1:2376 -v $PWD/.m2:/root/.m2 -v $PWD:/origin -v $PWD/dockerfiles/run/spring/destination:/destination -t james/project SHA1
```
Comment (mbaechler) [inline README.txt]: you should rather document the unix socket mount instead

File: mpt/impl/smtp/core/pom.xml [hunk 3]
```diff
@@ -156,6 +156,11 @@
                 </dependency>
                 <dependency>
                     <groupId>org.apache.james</groupId>
+                    <artifactId>james-server-util-java8</artifactId>
```
Comment (mbaechler) [inline mpt/impl/smtp/core/pom.xml]: what about ordering ?

File: mpt/pom.xml [hunk 4]
```diff
@@ -112,6 +112,7 @@
         <javax.mail.artifactId>mail</javax.mail.artifactId>
         <backends.version>1.8.3-SNAPSHOT</backends.version>
         <james.version>3.0.0-beta5-SNAPSHOT</james.version>
+        <testcontainers-version>1.0.2</testcontainers-version>
```
Comment (mbaechler) [inline mpt/pom.xml]: why do you want to extract the version ?

**Comment by mbaechler** [comment]
really cool

File: server/container/util-java8/src/test/java/org/apache/james/util/streams/SwarmGenericContainer.java (hunk 1 above)
Comment (rouazana) [inline server/container/util-java8/src/test/java/org/apache/james/util/streams/SwarmGenericContainer.java]: sorry, I thought the syntax was coherent with something like addEnv("affinity:container", "=value"); or addEnv("affinity:container", "!value"); As you need to write == and !=, it is ok for envVariables.add.

File: mpt/impl/smtp/core/pom.xml (hunk 3 above)
Comment (aduprat) [inline mpt/impl/smtp/core/pom.xml]: OK

File: mpt/pom.xml (hunk 4 above)
Comment (aduprat) [inline mpt/pom.xml]: OK

**Comment by chibenwa** [comment]
:+1:

**Comment by mbaechler** [comment]
:+1:

**Pull request merged**